### PR TITLE
Feature/aodv receiver

### PR DESCRIPTION
**Pull request by gustavosinbandera1**
This PR is in order to handle reception of packets with rfc5444 format

### CODE REVIEW

File: sys/include/net/gnrc/aodvv2/aodvv2.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ */
+void gnrc_aodvv2_init(void);
+
+/**
+ * @brief   Set the metric type. If metric_type does not match any known metric
+ *          types, no changes will be made.
+ *
+ * @param[in] metric_type       type of new metric
+ */
+void aodv_set_metric_type(aodvv2_metric_t metric_type);
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/aodvv2.h]:
```suggestion
void gnrc_aodvv2_set_metric_type(aodvv2_metric_t metric_type);
```

File: sys/include/net/gnrc/aodvv2/types.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+ * @author Locha Mesh Developers (contact@locha.io)
+ * @brief Main firmware file
+ * @version 0.1
+ * @date 2020-02-02
+ * 
+ * @copyright Copyright (c) 2020 Locha Mesh project developers
+ * @license Apache 2.0, see LICENSE file for details
+ * 
+ */
+
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h]: This can be removed.

File: sys/include/net/gnrc/aodvv2/types.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+/*
+ * Copyright (C) 2014 Freie Universität Berlin
+ * Copyright (C) 2014 Lotte Steenbrink <lotte.steenbrink@fu-berlin.de>
+ *
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h]:
```suggestion
 * Copyright (C) 2020 Locha Inc
 *
```

File: sys/include/net/gnrc/aodvv2/types.h [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * @ingroup     aodvv2
+ * @{
+ *
+ * @file        aodvv2/types.h
+ * @brief       data types for the aodvv2 routing protocol
+ *
+ * @author      Lotte Steenbrink <lotte.steenbrink@fu-berlin.de>
+ */
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h:20]:
```suggestion
 * @author      Gustavo Grisales <email@domain.gtld>
 */
```

File: sys/include/net/gnrc/aodvv2/types.h [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+#include <timex.h> //yo---
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   AODVv2 metric types. Extend to include alternate metrics.
+ */
+typedef enum {
+  HOP_COUNT = 3, /**< see RFC6551*/
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h]: Nit:
```suggestion
    HOP_COUNT = 3, /**< see RFC 6551 */
```

File: sys/include/net/gnrc/aodvv2/types.h [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+typedef uint16_t aodvv2_seqnum_t;
+
+#define AODVV2_DEFAULT_METRIC_TYPE HOP_COUNT
+
+/**
+ * @brief   AODVv2 message types
+ */
+enum rfc5444_msg_type {
+  RFC5444_MSGTYPE_RREQ = 10,
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h]:
```suggestion
  RFC5444_MSGTYPE_RREQ = 10, /**< RREQ message type */
```

File: sys/include/net/gnrc/aodvv2/types.h [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  RFC5444_MSGTYPE_RREP = 11,
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h]:
```suggestion
  RFC5444_MSGTYPE_RREP = 11, /**< RREP message type */
```

File: sys/include/net/gnrc/aodvv2/types.h [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  RFC5444_MSGTYPE_RERR = 12,
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h]:
```suggestion
  RFC5444_MSGTYPE_RERR = 12, /**< RERR message type */
```

File: sys/include/net/gnrc/aodvv2/types.h [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+};
+
+/**
+ * @brief   AODVv2 TLV types
+ */
+enum rfc5444_tlv_type {
+  RFC5444_MSGTLV_ORIGSEQNUM,
+  RFC5444_MSGTLV_TARGSEQNUM,
+  RFC5444_MSGTLV_UNREACHABLE_NODE_SEQNUM,
+  RFC5444_MSGTLV_METRIC,
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h]: Add documentation.

File: sys/net/gnrc/aodvv2/aodv.h [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+ * @author Locha Mesh Developers (contact@locha.io)
+ * @brief Main firmware file
+ * @version 0.1
+ * @date 2020-02-02
+ * 
+ * @copyright Copyright (c) 2020 Locha Mesh project developers
+ * @license Apache 2.0, see LICENSE file for details
+ * 
+ */
+
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/aodv.h]: Remove please.

File: sys/net/gnrc/aodvv2/constants.h [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+ * @author Locha Mesh Developers (contact@locha.io)
+ * @brief Main firmware file
+ * @version 0.1
+ * @date 2020-02-02
+ * 
+ * @copyright Copyright (c) 2020 Locha Mesh project developers
+ * @license Apache 2.0, see LICENSE file for details
+ * 
+ */
+
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/constants.h]: Remove please.

File: sys/net/gnrc/aodvv2/reader.c [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+
+#include "reader.h"
+
+
+
+#if VERBOSE_DEBUG
+#define VDEBUG(...) DEBUG(__VA_ARGS__)
+#else
+#define VDEBUG(...)
+#endif
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/reader.c]: You can remove this macro and use only the `DEBUG` macro.

File: sys/net/gnrc/aodvv2/reader.c [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+ * @author      Lotte Steenbrink <lotte.steenbrink@fu-berlin.de>
+ * @author Gustavo Grisales <gustavosinbandera1@hotmail.com>
+ */
+
+#ifdef RIOT
+#include "net_help.h"
+#endif
+#define ENABLE_DEBUG (1)
+
+#define VERBOSE_DEBUG (1)
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/reader.c]: Can be removed.

File: sys/net/gnrc/aodvv2/reader.h [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+ */
+/**
+                    
+ * @ingroup     aodvv2
+ *
+ * @file        aodv.c|
+ * @brief       aodvv2 routing protocol
+ *
+ * @author      Lotte Steenbrink <lotte.steenbrink@fu-berlin.de>
+ * @author Gustavo Grisales <gustavosinbandera1@hotmail.com>
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/reader.h]: Indent correctly.

File: sys/net/gnrc/aodvv2/utils.h [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+ * @file main.c
+ * @author Locha Mesh Developers (contact@locha.io)
+ * @brief Main firmware file
+ * @version 0.1
+ * @date 2020-02-02
+ * 
+ * @copyright Copyright (c) 2020 Locha Mesh project developers
+ * @license Apache 2.0, see LICENSE file for details
+ * 
+ */
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/utils.h]: Remove please.

File: sys/net/gnrc/aodvv2/writer.c [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+ * @author Locha Mesh Developers (contact@locha.io)
+ * @brief Main firmware file
+ * @version 0.1
+ * @date 2020-02-02
+ * 
+ * @copyright Copyright (c) 2020 Locha Mesh project developers
+ * @license Apache 2.0, see LICENSE file for details
+ * 
+ */
+
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/writer.c]: Remove please.

File: sys/net/gnrc/aodvv2/writer.h [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+ * @author Locha Mesh Developers (contact@locha.io)
+ * @brief Main firmware file
+ * @version 0.1
+ * @date 2020-02-02
+ * 
+ * @copyright Copyright (c) 2020 Locha Mesh project developers
+ * @license Apache 2.0, see LICENSE file for details
+ * 
+ */
+
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/writer.h]: Remove please.

**Review by jeandudey — CHANGES_REQUESTED** [review]
Some changes required.

File: sys/include/net/gnrc/aodvv2/aodvv2.h [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+ * @author Locha Mesh Developers (contact@locha.io)
+ * @brief Main firmware file
+ * @version 0.1
+ * @date 2020-02-02
+ * 
+ * @copyright Copyright (c) 2020 Locha Mesh project developers
+ * @license Apache 2.0, see LICENSE file for details
+ * 
+ */
+
```
Comment (luisan00) [inline sys/include/net/gnrc/aodvv2/aodvv2.h]: Delete this block

File: sys/net/gnrc/aodvv2/aodv.c [hunk 19]
```diff
@@ -0,0 +1,423 @@
...
+ * @author Locha Mesh Developers (contact@locha.io)
+ * @brief Main firmware file
+ * @version 0.1
+ * @date 2020-02-02
+ *
+ * @copyright Copyright (c) 2020 Locha Mesh project developers
+ * @license Apache 2.0, see LICENSE file for details
+ *
+ */
+
```
Comment (luisan00) [inline sys/net/gnrc/aodvv2/aodv.c]: Delete this block

File: sys/include/net/gnrc/aodvv2/types.h [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+  uint8_t hoplimit;           /**< Hop limit */
+  struct netaddr sender;      /**< IP address of the neighboring router
+                               *   which sent the RREQ/RREP*/
+  aodvv2_metric_t metricType; /**< Metric type */
+  struct node_data origNode;  /**< Data about the originating node */
+  struct node_data targNode;  /**< Data about the originating node */
+  timex_t timestamp;          /**< point at which the packet was (roughly)
+                               *   received. Note that this timestamp
+                               *   will be set after the packet has been
+                               *   successfully parsed. */
```
Comment (luisan00) [inline sys/include/net/gnrc/aodvv2/types.h]: Use inline the formula `/* */` for comments please

File: sys/net/gnrc/aodvv2/aodv.h [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+/**
+ * @brief   This struct contains data which needs to be put into a RREQ or RREP.
+ *          It is used to transport this data in a message to the sender_thread.
+ * @note    Please note that it is for internal use only. To send a RREQ or
+ * RREP, please use the aodv_send_rreq() and aodv_send_rrep() functions.
+ */
+struct rreq_rrep_data {
+  struct aodvv2_packet_data *packet_data; /**< Data for the RREQ or RREP */
+  struct netaddr *next_hop;               /**< Next hop to which the RREQ
+                                           *   or RREP should be sent */
```
Comment (luisan00) [inline sys/net/gnrc/aodvv2/aodv.h]: Use the inline `/*  */` formula

File: sys/net/gnrc/aodvv2/aodv.h [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+ *          please use the aodv_send_rerr() function.
+ */
+struct rerr_data {
+  struct unreachable_node
+      *unreachable_nodes;   /**< All unreachable nodes. Beware,
+                             *   this is the start of an array */
+  size_t len;               /**< Length of the unreachable_nodes array */
+  int hoplimit;             /**< hoplimit for the RERR */
+  struct netaddr *next_hop; /**< Next hop to which the RERR
+                             *   should be sent */
```
Comment (luisan00) [inline sys/net/gnrc/aodvv2/aodv.h]: Use the inline `/*  */` formula

File: sys/net/gnrc/aodvv2/writer.c [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+    [RFC5444_MSGTLV_METRIC] = {
+        .type = RFC5444_MSGTLV_METRIC,
+        .exttype = AODVV2_DEFAULT_METRIC_TYPE
+    },
+};
+
+
+
+
+
```
Comment (luisan00) [inline sys/net/gnrc/aodvv2/writer.c]: remove empty lines

File: sys/net/gnrc/aodvv2/writer.c [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+        return;
+    }
+
+    // /* Make sure no other thread is using the writer right now */
+    mutex_lock(&writer_mutex);
+    memcpy(&_target.packet_data, packet_data, sizeof(struct aodvv2_packet_data));
+    _target.type = RFC5444_MSGTYPE_RREQ;
+    _target.packet_data.hoplimit = packet_data->hoplimit;
+
+    // /* set address to which the write_packet callback should send our RREQ */
```
Comment (luisan00) [inline sys/net/gnrc/aodvv2/writer.c:167]: use the correct formula for to comment this, please

**Review by luisan00 — CHANGES_REQUESTED** [review]
Please review 🍡

File: sys/include/net/gnrc/aodvv2/aodvv2.h [hunk 25]
```diff
@@ -54,7 +42,7 @@ void gnrc_aodvv2_init(void);
  *
  * @param[in] metric_type       type of new metric
  */
-void aodv_set_metric_type(aodvv2_metric_t metric_type);
+void gnrc_aodv_set_metric_type(aodvv2_metric_t metric_type);
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/aodvv2.h]:
```suggestion
void gnrc_aodvv2_set_metric_type(aodvv2_metric_t metric_type);
```

File: sys/include/net/gnrc/aodvv2/types.h [hunk 26]
```diff
@@ -35,72 +25,79 @@
 #include "common/netaddr.h"
 #include <timex.h> //yo---
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h]:
```suggestion
#include <timex.h>
```

File: sys/include/net/gnrc/aodvv2/types.h [hunk 27]
```diff
@@ -35,72 +25,79 @@
...
 #ifdef __cplusplus
-extern "C" {
+extern "C"
+{
 #endif
 
-/**
+    /**
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h]:
```suggestion
/**
```

File: sys/include/net/gnrc/aodvv2/types.h [hunk 28]
```diff
@@ -35,72 +25,79 @@
...
  * @brief   AODVv2 metric types. Extend to include alternate metrics.
  */
-typedef enum {
-  HOP_COUNT = 3, /**< see RFC6551*/
-} aodvv2_metric_t;
+    typedef enum
+    {
+        HOP_COUNT = 3, /**< see RFC 6551*/
+    } aodvv2_metric_t;
```
Comment (jeandudey) [inline sys/include/net/gnrc/aodvv2/types.h]: Fix indentation.

File: sys/net/gnrc/aodvv2/aodv.c [hunk 29]
```diff
@@ -1,19 +1,7 @@
...
- * @version 0.1
- * @date 2020-02-02
- *
- * @copyright Copyright (c) 2020 Locha Mesh project developers
- * @license Apache 2.0, see LICENSE file for details
- *
- */
 
-/*
+/* Copyright (C) 2020 Locha Inc
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/aodv.c:2]: Put it down.

File: sys/net/gnrc/aodvv2/aodv.c [hunk 30]
```diff
@@ -93,331 +81,385 @@ static void *gnrc_aodvv2_sender_thread(void *arg);
...
+
+    wt = container_of(iface, struct writer_target, interface);
+
+    netaddr_to_ipv6_addr_t(&wt->target_addr, &temp_addr_dest);
+    DEBUG("!!!!!!!!!!!!!!!!!!!MIRA ESTA IP: %s",
+          ipv6_addr_to_str(addr_str, &temp_addr_dest, IPV6_ADDR_MAX_STR_LEN));
+
+    if (wt->type == RFC5444_MSGTYPE_RREQ)
+    {
+        DEBUG("<<<<<<<<<<<<<<<<<<<<>>>>>>>>>>>>>>>>>>>>ESTO ES UN ROUTE REQUEST\n");
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/aodv.c:347]:
```suggestion
        DEBUG("[_write_packet]: this is a RREQ.\n");
```

File: sys/net/gnrc/aodvv2/aodv.c [hunk 31]
```diff
@@ -93,331 +81,385 @@ static void *gnrc_aodvv2_sender_thread(void *arg);
...
+    if (sock_udp_create(&sock, &local, NULL, 0) < 0)
+    {
+        puts("Error creating UDP sock");
+        return 1;
+    }
 
-// receive RREQs, RREPs and RERRs and handle them
-static void *_aodv_receiver_thread(void *arg) {
-  (void)arg;
+    sock_udp_ep_t remote = {.family = AF_INET6};
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/aodv.c:378]:
```suggestion
    sock_udp_ep_t remote = {
        .family = AF_INET6,
        .port = MANET_PORT,
    };
```

File: sys/net/gnrc/aodvv2/aodv.c [hunk 32]
```diff
@@ -93,331 +81,385 @@ static void *gnrc_aodvv2_sender_thread(void *arg);
...
+    remote.port = MANET_PORT;
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/aodv.c:379]:
```suggestion
```

File: sys/net/gnrc/aodvv2/aodv.c [hunk 33]
```diff
@@ -93,331 +81,385 @@ static void *gnrc_aodvv2_sender_thread(void *arg);
...
-  //   }
-
-  return NULL;
+    DEBUG("Success: started UDP>>>to receive ROUTE MESSAGES");
+    while (1)
+    {
+        int res;
+        struct sockaddr_in6 src;
+        socklen_t src_len = sizeof(struct sockaddr_in6);
+        if ((res = recvfrom(_socket, buf_rcv, sizeof(buf_rcv), 0,
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/aodv.c:436]: Use `gnrc_sock_udp` API for this

File: sys/net/gnrc/aodvv2/aodv.h [hunk 34]
```diff
@@ -23,16 +12,16 @@
  * @ingroup     aodvv2
  * @{
  *
- * @file        aodv.h
- * @brief       aodvv2 routing protocol
+ * @file        aodvv2/types.h
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/aodv.h:15]:
```suggestion
 * @file        types.h
```

File: sys/net/gnrc/aodvv2/seqnum.c [hunk 35]
```diff
@@ -11,16 +12,15 @@
  * @ingroup     aodvv2
  * @{
  *
- * @file        seqnum.c
- * @brief       aodvv2 sequence number
+ * @file        aodvv2/types.h
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/seqnum.c:15]:
```suggestion
 * @file        seqnum.c
```

File: sys/net/gnrc/aodvv2/seqnum.c [hunk 36]
```diff
@@ -11,16 +12,15 @@
...
+ * @brief       data types for the aodvv2 routing protocol
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/seqnum.c:16]:
```suggestion
 * @brief       AODVv2 sequence number
```

File: sys/net/gnrc/aodvv2/utils.c [hunk 37]
```diff
@@ -11,10 +12,11 @@
  * @ingroup     aodvv2
  * @{
  *
- * @file        utils.c
- * @brief       client- and RREQ-table, ipv6 address representation converters
+ * @file        aodvv2/types.h
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/utils.c:15]:
```suggestion
 * @file        utils.c
```

File: sys/net/gnrc/aodvv2/utils.h [hunk 38]
```diff
@@ -22,10 +12,11 @@
  * @ingroup     aodvv2
  * @{
  *
- * @file        utils.h
- * @brief       client- and RREQ-table, ipv6 address representation converters
+ * @file        aodvv2/types.h
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/utils.h:15]:
```suggestion
 * @file        utils.h
```

File: sys/net/gnrc/aodvv2/writer.h [hunk 39]
```diff
@@ -23,10 +12,11 @@
  * @ingroup     aodvv2
  * @{
  *
- * @file        writer.h
- * @brief       writer to create RFC5444 aodvv2 messages
+ * @file        aodvv2/types.h
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/writer.h:15]:
```suggestion
 * @file        writer.c
```

File: sys/net/gnrc/aodvv2/writer.h [hunk 40]
```diff
@@ -23,10 +12,11 @@
...
+ * @brief       data types for the aodvv2 routing protocol
```
Comment (jeandudey) [inline sys/net/gnrc/aodvv2/writer.h:16]:
```suggestion
 * @brief       Writer to create RFC 5444 AODVv2 messages
```

**Review by jeandudey — CHANGES_REQUESTED** [review]

File: sys/include/net/gnrc/aodvv2/aodvv2.h (hunk 25 above)
Comment (jeandudey) [reply sys/include/net/gnrc/aodvv2/aodvv2.h]: @gustavosinbandera1 not done yet.

File: sys/include/net/gnrc/aodvv2/types.h [hunk 41]
```diff
@@ -0,0 +1,106 @@
...
+
+/**
+ * @ingroup     aodvv2
+ * @{
+ *
+ * @file        aodvv2/types.h
+ * @brief       data types for the aodvv2 routing protocol
+ *
+ * @author      Lotte Steenbrink <lotte.steenbrink@fu-berlin.de>
+ * @author      Gustavo Grisales <gustavosinbandera1@hotmail.com.com>
```
Comment (luisan00) [inline sys/include/net/gnrc/aodvv2/types.h:19]: change gustavosinbandera1@hotmail.com.com to gustavosinbandera1@hotmail.com

File: sys/net/gnrc/aodvv2/aodv.c [hunk 42]
```diff
@@ -0,0 +1,465 @@
...
+
+/**
+ * @ingroup     aodvv2
+ * @{
+ *
+ * @file        aodvv2/types.h
+ * @brief       data types for the aodvv2 routing protocol
+ *
+ * @author      Lotte Steenbrink <lotte.steenbrink@fu-berlin.de>
+ * @author      Gustavo Grisales <gustavosinbandera1@hotmail.com.com>
```
Comment (luisan00) [inline sys/net/gnrc/aodvv2/aodv.c:19]: change gustavosinbandera1@hotmail.com.com to gustavosinbandera1@hotmail.com

File: sys/net/gnrc/aodvv2/aodv.c [hunk 43]
```diff
@@ -0,0 +1,465 @@
...
+        DEBUG("unspecified address\n");
+        return (ipv6_addr_t)IPV6_ADDR_UNSPECIFIED;
+    }
+
+    for (unsigned i = 0; i < (unsigned)(r / sizeof(ipv6_addr_t)); i++)
+    {
+        char ipv6_address[IPV6_ADDR_MAX_STR_LEN];
+        ipv6_addr_to_str(ipv6_address, &ipv6_addr, IPV6_ADDR_MAX_STR_LEN);
+    }
+                                                                                return ipv6_addr;
```
Comment (luisan00) [inline sys/net/gnrc/aodvv2/aodv.c:315]: Fix Indentation

**Review by luisan00 — CHANGES_REQUESTED** [review]
🍔

**Comment by jeandudey** [comment]
This work was merged on btcven/radio-firmware already, with modifications from @gustavosinbandera1 and my person.

**Pull request closed without merging**